docs/client-design.md

```diff
@@ -0,0 +1,287 @@
+    
+    /** Needs to be called by the application on a read event. */
+    void read_ready();
+    /** Needs to be called by the application on a write event. */
+    void write_ready();
+};
+```
+
+### Connections
+
+```c++
+/**
+ * A connection encapsulates sending requests and receiving responses from one 
+ * Tarantool instance.
+ */
+template<class IOEventProvider>
+class Connection {
+public:
+    Connection(IOEventProvider &net_provider, 
+	           const ConnectionOptions &connection_options);
+	
+    /** 
+	 * Return the connection's status.
+	 */
+	enum class ConnectionStatus get_status() const;
+	
+	/** 
+	 * If the connection is an erroneous state (see `status`), return the 
+	 * connection error that caused it. The same error is also passed to 
+	 * response callbacks, and the same error is returned by 
+	 * `Request::get_error`. 
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/* An abstract request's interface. A request object is always returned. */
+	Request some_request(/* options */);
+};
+```
+
+#### Connect Options
+
+```c++
+/* Extend `ConnectionOptions` with an option for the reconnection feature. */
+struct ConnectOptions {
+	/* All existing options. */
+	
+	/** 
+	 * In the event of a broken connection, the interval in which the stream 
+	 * tries to re-establish the connection. 
+	 */
+	static constexpr size_t DEFAULT_RECONNECTION_INTERVAL = 2;
+	size_t reconnection_interval = DEFAULT_RECONNECTION_INTERVAL;
+};
+```
+
+### Requests
+
+```c++
+/** Encapsulates management of a request issued through a connection. */
+class Request {
+public:
+	/** 
+	 * Callback called when response is ready. Since the callback has a fixed 
+	 * signature, we need to allow for capturing additional context using 
+	 * lambdas. Hence, we use `std::function` for type erasure.
+	 * 
+	 * See `Connection::get_error` for details about the `error` parameter.
+	 */ 
+	using request_cb_f = 
+		std::function<void(const std::optional<ConnectionError> &error, 
+		                   Response &&response, Connection &connection)>;
+	
+	/** Set a callback called when the response is ready. */
+	int set_callback(request_cb_f request_cb);
+	
+	/** Remove the callback associated with the request, if any. */
+	void reset_callback();
+	
+	/** Return the request's status. */
+	enum class RequestStatus get_status();
+	
+	/**
+	 * Return a connection error, if any. See `Connection::get_error` for 
+	 * details.
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/**
+	 * Cancel the request, ending the lifetime of the corresponding response.
+	 */
+	void cancel();
+	
+	/** 
+	 * Return the response, if any. A response is available, iff:
+	 * 1. The response has been received by the connection.
+	 * 2. The response has not been dispatched to a callback.
+	 * 3. The response has not already been retrieved previously.
+	 */
+	std::optional<Response> get_response();
+};
```

Comment:
We could create extract-once class:
1. add `isReady` method
2. `set_callback` should be rvalue method
3. `get_response` -> `get_result` (`result` is actually `variant<Error, Response>`) - rvalue method
4. `reset_callback` will be dropped. Anyway, I don't see any use cases for it.

Then, user cannot get response after callback is set and vice versa.

Reply:
> 1. add `isReady` method

I don't understand, why we do need this method, if we have `get_status`, which we could check for request readiness?

> 2. `set_callback` should be rvalue method

This sounds like a cool solution. I suggest we keep the request object and pass it to the callback later on.

> 3. `get_response` -> `get_result` (`result` is actually `variant<Error, Response>`) - rvalue method

I don't like this, since it would be nice to share the `ConnectionError` object between the connection and the requests (a variant does not allow to store references).

> 4. `reset_callback` will be dropped. Anyway, I don't see any use cases for it.

Initially, I thought there might be some conditions when an application would want to remove the callback. But now I come to think that it is indeed redundant.